File: 문자열처리/KAKAO2018_newclurstering.js
function solution(str1, str2) {
  var answer = 0;
  const multiSet1 = makeMultiSet(str1)
  const multiSet2 = makeMultiSet(str2) //다중집합 1,2
  const inter = []
  let multiSet2_copy = [...multiSet2]  //[...multiSet2] 
  multiSet1.forEach((word)=>{
      if(multiSet2_copy.includes(word)){
          inter.push(word)        
          multiSet2_copy.splice(multiSet2_copy.indexOf(word),1)
          // 원본에서 해당 인덱스의 문자 하나 지우고기
          // 틀린코드 : multiSet2_copy = multiSet2_copy.filter((x)=>x!==word)
          //filter는 원본을 직접 변경하지도 않고, 중복제거가 이뤄지지 않은 채 새로운 배열을 반환함 따라서 원본변경 메서드인 splice로 직접 제거해야함!
      }
  }) // 교집합 forEach는 배열에 어떤 로직을 수행만 할때
  let len_inter = inter.length
  let len_union = multiSet2.length + multiSet1.length - inter.length
  console.log(len_inter,len_union)
  // 중복포함 다중집합이어도 합집합의 갯수는 일반집합의 공식과 같다.
  if (len_inter === 0 && len_union ===0){
      return 1 * 65536
  }
  else{
      return Math.floor((len_inter / len_union)*65536)
  }
}
function makeMultiSet(str){
      str = str.toLowerCase()
      let multiSet = []
      for (let i = 0; i<str.length-1; i++){
          let tempo = str.slice(i,i+2)
          if ( /[a-z]{2}/.test(tempo) ){
              multiSet.push(tempo)//2개
      }
  }
  return multiSet
}

File: 문자열처리/KAKAO2018_newclurstering.py
import re 
import math
def solution(str1, str2):
    p = re.compile('[a-z]{2}') # isalpha()로 대체 가능
    def makeMulti(str):
        multiSet = []
        str = str.lower()
        for i in range(len(str)-1):
            word = str[i:i+2]
            if p.search(word) and p.match(word).group() :
                multiSet.append(word)
        return multiSet
    multi1 = makeMulti(str1)
    multi2 = makeMulti(str2)
    def jakardCalculation(mulit1=multi1,multi2=multi2):
        inter = []
        multi2_copy = multi2.copy() # 원본은 그대로 남겨둬야함 
        for word in mulit1:
            if word in multi2_copy:
                inter.append(word)
                multi2_copy.remove(word) # 교집합을 구하며 처리한 요소는 삭제필요
        len_inter = len(inter)
        len_union = len(multi1+multi2) - len(inter)
        if len_inter == 0 and len_union ==0:
            return 1
        else:
            return len_inter / len_union
    return math.floor(jakardCalculation()*65536) # int()로 대체 가능 
                
        
  